File: SMLite.Test/SMLite.Test.cpp
#include "CppUnitTest.h"
#include "../SMLite/SMLite.hpp"

#include <sstream>

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

enum class MyState { Rest, Ready, Reading, Writing };
enum class MyTrigger { Run, Close, Read, FinishRead, Write, FinishWrite };



std::wstring _wstr (MyState _state) {
	std::wstringstream _ss;
	_ss << (int) _state;
	std::wstring _ret;
	_ss >> _ret;
	return _ret;
}

std::string _append (std::string _p1, int _p2) {
	std::stringstream _ss;
	_ss << _p1 << _p2;
	std::string _ret;
	_ss >> _ret;
	return _ret;
}

namespace Microsoft {
	namespace VisualStudio {
		namespace CppUnitTestFramework {
			template<> static std::wstring ToString<MyState> (const MyState &t) { return _wstr (t); }
			template<> static std::wstring ToString<MyState> (const MyState *t) { return _wstr (*t); }
			template<> static std::wstring ToString<MyState> (MyState *t) { return _wstr (*t); }
		}
	}
}



namespace SMLiteTest {
	TEST_CLASS (SMLiteTest) {
	public:
		TEST_METHOD (TestMethod1) {
			int n = 0;
			bool entry_one = true;
			Fawdlstty::SMLiteBuilder<MyState, MyTrigger> _smb {};
			_smb.Configure (MyState::Rest)
				->OnEntry ([&] () { Assert::IsFalse (entry_one); entry_one = true; n += 1; })
				->OnLeave ([&] () { Assert::IsTrue (entry_one); entry_one = false; n += 10; })
				->WhenChangeTo (MyTrigger::Run, MyState::Ready)
				->WhenIgnore (MyTrigger::Close);
			_smb.Configure (MyState::Ready)
				->OnEntry ([&] () { Assert::IsFalse (entry_one); entry_one = true; n += 100; })
				->OnLeave ([&] () { Assert::IsTrue (entry_one); entry_one = false; n += 1000; })
				->WhenChangeTo (MyTrigger::Read, MyState::Reading)
				->WhenChangeTo (MyTrigger::Write, MyState::Writing)
				->WhenChangeTo (MyTrigger::Close, MyState::Rest);
			_smb.Configure (MyState::Reading)
				->OnEntry ([&] () { Assert::IsFalse (entry_one); entry_one = true; n += 10000; })
				->OnLeave ([&] () { Assert::IsTrue (entry_one); entry_one = false; n += 100000; })
				->WhenChangeTo (MyTrigger::FinishRead, MyState::Ready)
				->WhenChangeTo (MyTrigger::Close, MyState::Rest);
			_smb.Configure (MyState::Writing)
				->OnEntry ([&] () { Assert::IsFalse (entry_one); entry_one = true; n += 1000000; })
				->OnLeave ([&] () { Assert::IsTrue (entry_one); entry_one = false; n += 10000000; })
				->WhenChangeTo (MyTrigger::FinishWrite, MyState::Ready)
				->WhenChangeTo (MyTrigger::Close, MyState::Rest);

			auto _sm = _smb.Build (MyState::Rest);
			Assert::AreEqual (_sm->GetState (), MyState::Rest);
			Assert::AreEqual (n, 0);
			Assert::IsTrue (_sm->AllowTriggering (MyTrigger::Run));
			Assert::IsTrue (_sm->AllowTriggering (MyTrigger::Close));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::Read));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::FinishRead));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::Write));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::FinishWrite));

			_sm->Triggering (MyTrigger::Close);
			Assert::AreEqual (_sm->GetState (), MyState::Rest);
			Assert::AreEqual (n, 0);

			_sm->Triggering (MyTrigger::Close);
			Assert::AreEqual (_sm->GetState (), MyState::Rest);
			Assert::AreEqual (n, 0);

			_sm->Triggering (MyTrigger::Close);
			Assert::AreEqual (_sm->GetState (), MyState::Rest);
			Assert::AreEqual (n, 0);

			_sm->Triggering (MyTrigger::Run);
			Assert::AreEqual (_sm->GetState (), MyState::Ready);
			Assert::AreEqual (n, 110);
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::Run));
			Assert::IsTrue (_sm->AllowTriggering (MyTrigger::Close));
			Assert::IsTrue (_sm->AllowTriggering (MyTrigger::Read));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::FinishRead));
			Assert::IsTrue (_sm->AllowTriggering (MyTrigger::Write));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::FinishWrite));

			_sm->Triggering (MyTrigger::Close);
			Assert::AreEqual (_sm->GetState (), MyState::Rest);
			Assert::AreEqual (n, 1111);
			Assert::IsTrue (_sm->AllowTriggering (MyTrigger::Run));
			Assert::IsTrue (_sm->AllowTriggering (MyTrigger::Close));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::Read));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::FinishRead));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::Write));
			Assert::IsFalse (_sm->AllowTriggering (MyTrigger::FinishWrite));

			_sm->Triggering (MyTrigger::Run);
			Assert::AreEqual (_sm->GetState (), MyState::Ready);
			Assert::AreEqual (n, 1221);

			_sm->Triggering (MyTrigger::Read);
			Assert::AreEqual (_sm->GetState (), MyState::Reading);
			Assert::AreEqual (n, 12221);

			_sm->Triggering (MyTrigger::FinishRead);
			Assert::AreEqual (_sm->GetState (), MyState::Ready);
			Assert::AreEqual (n, 112321);

			_sm->Triggering (MyTrigger::Write);
			Assert::AreEqual (_sm->GetState (), MyState::Writing);
			Assert::AreEqual (n, 1113321);

			_sm->Triggering (MyTrigger::FinishWrite);
			Assert::AreEqual (_sm->GetState (), MyState::Ready);
			Assert::AreEqual (n, 11113421);

			_sm->SetState (MyState::Reading);
			_sm->SetState (MyState::Writing);
			_sm->SetState (MyState::Rest);
			Assert::AreEqual (n, 11113421);
		}

		TEST_METHOD (TestMethod3) {
			std::string s = "";
			std::function<void (MyState, MyTrigger, std::string)> _f = [&] (MyState _state, MyTrigger _trigger, std::string _p1) { s = _p1; };
			Fawdlstty::SMLiteBuilder<MyState, MyTrigger> _smb {};
			_smb.Configure (MyState::Rest)
				->WhenFunc (MyTrigger::Run, std::function<MyState (MyState, MyTrigger)> (
					[&] (MyState _state, MyTrigger _trigger) { s = "WhenFunc_Run"; return MyState::Ready; }))
				->WhenFunc (MyTrigger::Read, std::function<MyState (MyState, MyTrigger, std::string)> (
					[&] (MyState _state, MyTrigger _trigger, std::string _p1) { s = _p1; return MyState::Ready; }))
				->WhenFunc (MyTrigger::FinishRead, std::function<MyState (MyState, MyTrigger, std::string, int)> (
					[&] (MyState _state, MyTrigger _trigger, std::string _p1, int _p2) { s = _append (_p1, _p2); return MyState::Ready; }))
				->WhenAction (MyTrigger::Close, std::function<void (MyState, MyTrigger)> (
					[&] (MyState _state, MyTrigger _trigger) { s = "WhenAction_Close"; }))
				->WhenAction (MyTrigger::Write, std::function<void (MyState, MyTrigger, std::string)> (
					[&] (MyState _state, MyTrigger _trigger, std::string _p1) { s = _p1; }))
				->WhenAction (MyTrigger::FinishWrite, std::function<void (MyState, MyTrigger, std::string, int)> (
					[&] (MyState _state, MyTrigger _trigger, std::string _p1, int _p2) { s = _append (_p1, _p2); }));
			_smb.Configure (MyState::Ready)
				->WhenChangeTo (MyTrigger::Close, MyState::Rest);

			auto _sm = _smb.Build (MyState::Rest);
			Assert::AreEqual (_sm->GetState (), MyState::Rest);
			Assert::AreEqual (s, std::string (""));

			_sm->Triggering (MyTrigger::Run);
			Assert::AreEqual (_sm->GetState (), MyState::Ready);
			Assert::AreEqual (s, std::string ("WhenFunc_Run"));
			_sm->Triggering (MyTrigger::Close);

			_sm->Triggering (MyTrigger::Read, std::string ("hello"));
			Assert::AreEqual (_sm->GetState (), MyState::Ready);
			Assert::AreEqual (s, std::string ("hello"));
			_sm->Triggering (MyTrigger::Close);

			_sm->Triggering (MyTrigger::FinishRead, std::string ("hello"), 1);
			Assert::AreEqual (_sm->GetState (), MyState::Ready);
			Assert::AreEqual (s, std::string ("hello1"));
			_sm->Triggering (MyTrigger::Close);

			_sm->Triggering (MyTrigger::Close);
			Assert::AreEqual (s, std::string ("WhenAction_Close"));
			Assert::AreEqual (_sm->GetState (), MyState::Rest);

			_sm->Triggering (MyTrigger::Write, std::string ("world"));
			Assert::AreEqual (s, std::string ("world"));
			Assert::AreEqual (_sm->GetState (), MyState::Rest);

			_sm->Triggering (MyTrigger::FinishWrite, std::string ("world"), 1);
			Assert::AreEqual (s, std::string ("world1"));
			Assert::AreEqual (_sm->GetState (), MyState::Rest);
		}
	};
}

File: src_python/SMLite.Test/SMLite.Test.py
# -*- coding: utf-8 -*-

import os
import sys
from enum import IntEnum
from SMLite import SMLite
from SMLiteBuilder import SMLiteBuilder

class MyState (IntEnum):
	Rest = 0
	Ready = 1
	Reading = 2
	Writing = 3

class MyTrigger (IntEnum):
	Run = 0
	Close = 1
	Read = 2
	FinishRead = 3
	Write = 4
	FinishWrite = 5



class Assert (object):
	@staticmethod
	def IsTrue (val):
		if not val:
			raise Exception ("Assert Failed")

	@staticmethod
	def IsFalse (val):
		if val:
			raise Exception ("Assert Failed")

	@staticmethod
	def AreEqual (val1, val2):
		if val1 != val2:
			raise Exception ("Assert Failed")



n = 0
entry_one = True
s = ""

def TestMethod1 ():
	def _rest_entry ():
		global n, entry_one
		Assert.IsTrue (not entry_one)
		entry_one = True
		n += 1
	def _rest_leave ():
		global n, entry_one
		Assert.IsTrue (entry_one)
		entry_one = False
		n += 10
	def _ready_entry ():
		global n, entry_one
		Assert.IsTrue (not entry_one)
		entry_one = True
		n += 100
	def _ready_leave ():
		global n, entry_one
		Assert.IsTrue (entry_one)
		entry_one = False
		n += 1000
	def _reading_entry ():
		global n, entry_one
		Assert.IsTrue (not entry_one)
		entry_one = True
		n += 10000
	def _reading_leave ():
		global n, entry_one
		Assert.IsTrue (entry_one)
		entry_one = False
		n += 100000
	def _writing_entry ():
		global n, entry_one
		Assert.IsTrue (not entry_one)
		entry_one = True
		n += 1000000
	def _writing_leave ():
		global n, entry_one
		Assert.IsTrue (entry_one)
		entry_one = False
		n += 10000000

	_smb = SMLiteBuilder ()
	_smb.Configure (MyState.Rest)\
		.OnEntry (_rest_entry)\
		.OnLeave (_rest_leave)\
		.WhenChangeTo (MyTrigger.Run, MyState.Ready)\
		.WhenIgnore (MyTrigger.Close)
	_smb.Configure (MyState.Ready)\
		.OnEntry (_ready_entry)\
		.OnLeave (_ready_leave)\
		.WhenChangeTo (MyTrigger.Read, MyState.Reading)\
		.WhenChangeTo (MyTrigger.Write, MyState.Writing)\
		.WhenChangeTo (MyTrigger.Close, MyState.Rest)
	_smb.Configure (MyState.Reading)\
		.OnEntry (_reading_entry)\
		.OnLeave (_reading_leave)\
		.WhenChangeTo (MyTrigger.FinishRead, MyState.Ready)\
		.WhenChangeTo (MyTrigger.Close, MyState.Rest)
	_smb.Configure (MyState.Writing)\
		.OnEntry (_writing_entry)\
		.OnLeave (_writing_leave)\
		.WhenChangeTo (MyTrigger.FinishWrite, MyState.Ready)\
		.WhenChangeTo (MyTrigger.Close, MyState.Rest)

	_sm = _smb.Build (MyState.Rest)
	Assert.AreEqual (_sm.GetState (), MyState.Rest)
	Assert.AreEqual (n, 0)
	Assert.IsTrue (_sm.AllowTriggering (MyTrigger.Run))
	Assert.IsTrue (_sm.AllowTriggering (MyTrigger.Close))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.Read))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.FinishRead))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.Write))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.FinishWrite))

	_sm.Triggering (MyTrigger.Close)
	Assert.AreEqual (_sm.GetState (), MyState.Rest)
	Assert.AreEqual (n, 0)

	_sm.Triggering (MyTrigger.Close)
	Assert.AreEqual (_sm.GetState (), MyState.Rest)
	Assert.AreEqual (n, 0)

	_sm.Triggering (MyTrigger.Close)
	Assert.AreEqual (_sm.GetState (), MyState.Rest)
	Assert.AreEqual (n, 0)

	_sm.Triggering (MyTrigger.Run)
	Assert.AreEqual (_sm.GetState (), MyState.Ready)
	Assert.AreEqual (n, 110)
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.Run))
	Assert.IsTrue (_sm.AllowTriggering (MyTrigger.Close))
	Assert.IsTrue (_sm.AllowTriggering (MyTrigger.Read))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.FinishRead))
	Assert.IsTrue (_sm.AllowTriggering (MyTrigger.Write))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.FinishWrite))

	_sm.Triggering (MyTrigger.Close)
	Assert.AreEqual (_sm.GetState (), MyState.Rest)
	Assert.AreEqual (n, 1111)
	Assert.IsTrue (_sm.AllowTriggering (MyTrigger.Run))
	Assert.IsTrue (_sm.AllowTriggering (MyTrigger.Close))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.Read))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.FinishRead))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.Write))
	Assert.IsFalse (_sm.AllowTriggering (MyTrigger.FinishWrite))

	_sm.Triggering (MyTrigger.Run)
	Assert.AreEqual (_sm.GetState (), MyState.Ready)
	Assert.AreEqual (n, 1221)

	_sm.Triggering (MyTrigger.Read)
	Assert.AreEqual (_sm.GetState (), MyState.Reading)
	Assert.AreEqual (n, 12221)

	_sm.Triggering (MyTrigger.FinishRead)
	Assert.AreEqual (_sm.GetState (), MyState.Ready)
	Assert.AreEqual (n, 112321)

	_sm.Triggering (MyTrigger.Write)
	Assert.AreEqual (_sm.GetState (), MyState.Writing)
	Assert.AreEqual (n, 1113321)

	_sm.Triggering (MyTrigger.FinishWrite)
	Assert.AreEqual (_sm.GetState (), MyState.Ready)
	Assert.AreEqual (n, 11113421)

	_sm.SetState (MyState.Reading)
	_sm.SetState (MyState.Writing)
	_sm.SetState (MyState.Rest)
	Assert.AreEqual (n, 11113421)

def TestMethod3 ():
	def _rest_run (_state, _trigger):
		global s
		s = "WhenFunc_Run"
		return MyState.Ready
	def _rest_read (_state, _trigger, _p1):
		global s
		s = _p1
		return MyState.Ready
	def _rest_finishread (_state, _trigger, _p1, _p2):
		global s
		s = _p1 + str (_p2)
		return MyState.Ready
	def _rest_close (_state, _trigger):
		global s
		s = "WhenAction_Close"
	def _rest_write (_state, _trigger, _p1):
		global s
		s = _p1
	def _rest_finishwrite (_state, _trigger, _p1, _p2):
		global s
		s = _p1 + str (_p2)

	_smb = SMLiteBuilder ()
	_smb.Configure (MyState.Rest)\
		.WhenFunc (MyTrigger.Run, _rest_run)\
		.WhenFunc (MyTrigger.Read, _rest_read)\
		.WhenFunc (MyTrigger.FinishRead, _rest_finishread)\
		.WhenAction (MyTrigger.Close, _rest_close)\
		.WhenAction (MyTrigger.Write, _rest_write)\
		.WhenAction (MyTrigger.FinishWrite, _rest_finishwrite)
	_smb.Configure (MyState.Ready)\
		.WhenChangeTo (MyTrigger.Close, MyState.Rest)

	_sm = _smb.Build (MyState.Rest)
	Assert.AreEqual (_sm.GetState (), MyState.Rest)
	Assert.AreEqual (s, "")

	_sm.Triggering (MyTrigger.Run)
	Assert.AreEqual (_sm.GetState (), MyState.Ready)
	Assert.AreEqual (s, "WhenFunc_Run")
	_sm.Triggering (MyTrigger.Close)

	_sm.Triggering (MyTrigger.Read, "hello")
	Assert.AreEqual (_sm.GetState (), MyState.Ready)
	Assert.AreEqual (s, "hello")
	_sm.Triggering (MyTrigger.Close)

	_sm.Triggering (MyTrigger.FinishRead, "hello", 1)
	Assert.AreEqual (_sm.GetState (), MyState.Ready)
	Assert.AreEqual (s, "hello1")
	_sm.Triggering (MyTrigger.Close)

	_sm.Triggering (MyTrigger.Close)
	Assert.AreEqual (s, "WhenAction_Close")
	Assert.AreEqual (_sm.GetState (), MyState.Rest)

	_sm.Triggering (MyTrigger.Write, "world")
	Assert.AreEqual (s, "world")
	Assert.AreEqual (_sm.GetState (), MyState.Rest)

	_sm.Triggering (MyTrigger.FinishWrite, "world", 1)
	Assert.AreEqual (s, "world1")
	Assert.AreEqual (_sm.GetState (), MyState.Rest)

if __name__ == '__main__':
	#lib_path = os.path.abspath (os.path.join ("../SMLite"))
	#sys.path.append (lib_path)
	#import SMLite
	#import SMLite.SMLiteBuilder
	TestMethod1 ()
	TestMethod3 ()
	input ("Test success, press ENTER to exit.")
